### PR TITLE
Fix lr scheduler

### PR DESCRIPTION
**Pull request by CarlosGomes98**
Addresses #1213, whose root cause was an arithmetic error in the lr scheduler.

The lr_scheduler adds 1 to the current step, which is not necessary and causes several errors in the lr steps.

Also took the chance to slightly simplify the logic (in my opinion) using early returns, as well as explicitly dealing with the edge case where decay_ratio is 0.

Copying over the reasoning I shared in #1213 

Lets set an example scenario where our lr schedule has warmup_steps of 5 and a target lr of 0.0008.
We would expect the lr to increase in each step by `0.0008/5 = 0.00016`. However, what happens is below:

```
[rank0]:[titan] 2025-06-04 14:42:54,945 - root - INFO - Step 1 lr: 0.00013333333333333334
[rank0]:[titan] 2025-06-04 14:42:54,946 - root - INFO - step:  1  loss:  1.7905  memory: 19.94GiB(42.08%)  tps: 519,779  tflops: 0.00  mfu: 0.00%
[rank0]:[titan] 2025-06-04 14:42:54,946 - root - INFO - Synchronizing and adjusting timeout for all ProcessGroups to 0:01:40
[rank0]:[titan] 2025-06-04 14:42:55,316 - root - INFO - Step 2 lr: 0.0002666666666666667
[rank0]:[titan] 2025-06-04 14:42:55,316 - root - INFO - step:  2  loss:  1.9953  memory: 20.45GiB(43.17%)  tps: 2,121,432  tflops: 0.00  mfu: 0.00%
[rank0]:[titan] 2025-06-04 14:42:55,686 - root - INFO - Step 3 lr: 0.0004
[rank0]:[titan] 2025-06-04 14:42:55,686 - root - INFO - step:  3  loss:  1.7081  memory: 20.45GiB(43.17%)  tps: 2,125,984  tflops: 0.00  mfu: 0.00%
[rank0]:[titan] 2025-06-04 14:42:56,079 - root - INFO - Step 4 lr: 0.0005333333333333334
[rank0]:[titan] 2025-06-04 14:42:56,079 - root - INFO - step:  4  loss:  1.6267  memory: 20.45GiB(43.17%)  tps: 2,004,664  tflops: 0.00  mfu: 0.00%
[rank0]:[titan] 2025-06-04 14:42:56,476 - root - INFO - Step 5 lr: 0.0008
[rank0]:[titan] 2025-06-04 14:42:56,477 - root - INFO - step:  5  loss:  1.5076  memory: 20.45GiB(43.17%)  tps: 1,977,805  tflops: 0.00  mfu: 0.00%
```

This is clearly wrong, as the lr is increasing by 0.00013 except for the last step, which increases by more to make up the difference.

We can easily resolve this by removing the assumptions in the comments `# 0-indexed step, hence + 1 adjustments`.
Making those changes, we get the desired behaviour of:

```
[rank0]:[titan] 2025-06-04 14:46:48,496 - root - INFO - Step 1 lr: 0.00016
[rank0]:[titan] 2025-06-04 14:46:48,497 - root - INFO - step:  1  loss:  1.8309  memory: 19.94GiB(42.08%)  tps: 512,095  tflops: 0.00  mfu: 0.00%
[rank0]:[titan] 2025-06-04 14:46:48,497 - root - INFO - Synchronizing and adjusting timeout for all ProcessGroups to 0:01:40
[rank0]:[titan] 2025-06-04 14:46:48,867 - root - INFO - Step 2 lr: 0.00032
[rank0]:[titan] 2025-06-04 14:46:48,867 - root - INFO - step:  2  loss:  1.9606  memory: 20.45GiB(43.17%)  tps: 2,124,707  tflops: 0.00  mfu: 0.00%
[rank0]:[titan] 2025-06-04 14:46:49,233 - root - INFO - Step 3 lr: 0.00048
[rank0]:[titan] 2025-06-04 14:46:49,233 - root - INFO - step:  3  loss:  1.6091  memory: 20.45GiB(43.17%)  tps: 2,147,417  tflops: 0.00  mfu: 0.00%
[rank0]:[titan] 2025-06-04 14:46:49,640 - root - INFO - Step 4 lr: 0.00064
[rank0]:[titan] 2025-06-04 14:46:49,641 - root - INFO - step:  4  loss:  1.4623  memory: 20.45GiB(43.17%)  tps: 1,931,065  tflops: 0.00  mfu: 0.00%
[rank0]:[titan] 2025-06-04 14:46:50,039 - root - INFO - Step 5 lr: 0.0008
[rank0]:[titan] 2025-06-04 14:46:50,040 - root - INFO - step:  5  loss:  1.6065  memory: 20.45GiB(43.17%)  tps: 1,971,288  tflops: 0.00  mfu: 0.00%
```

This same behaviour was causing the bug I described in #1213 . It was making it so, in the final step, we were going 1 over the max_steps expected, and thus setting the multiplicative lr adjustment factor to 0.

### CODE REVIEW

File: torchtitan/components/lr_scheduler.py [hunk 1]
```diff
@@ -142,25 +142,26 @@ def linear_warmup_stable_decay(
         to ensure the learning rate does not drop below this minimum value.
         """
         warmup_stable_steps = warmup_steps + stable_steps
+        # if we are in the warmup phase, return the warmup progress
+        # if warmup_steps is 0, we will go to the next phase
         if current_step < warmup_steps:
```
Comment (wwwjn) [inline torchtitan/components/lr_scheduler.py:153]: Nice fix, thank you @CarlosGomes98 ! I agree `warmup_steps` should not +1, and it's not 0-indexed. For `current_step`, because we always call `self.step+=1 ` at the beginning of train(), `current_step` is also not 0-indexed. (https://github.com/pytorch/torchtitan/blob/refs/heads/main/torchtitan/train.py#L434)

**Review by fegin — CHANGES_REQUESTED** [review]
There was a discussion in `https://github.com/pytorch/torchtitan/pull/1010`. I'm wondering should we have unittests to verify different use cases for `linear_warmup_stable_decay()`? This function is a pure Python function which should not be hard to add some test cases to see if we fixes both issues (the issue from the checkpoint and the issue from the previous divided by zero).

File: torchtitan/components/lr_scheduler.py [hunk 2]
```diff
@@ -142,25 +142,26 @@ def linear_warmup_stable_decay(
...
-            progress = float(current_step - warmup_stable_steps) / (decay_steps + 1)
-
-            if lr_decay_type == "linear":
-                curr_adjustment = 1 - progress
-            elif lr_decay_type == "sqrt":
-                curr_adjustment = 1 - math.sqrt(progress)
-            elif lr_decay_type == "cosine":
-                curr_adjustment = 0.5 * (1.0 + math.cos(math.pi * progress))
-            curr_adjustment = lr_min + (1 - lr_min) * curr_adjustment
+            return float(current_step / warmup_steps)
```
Comment (tianyu-l) [inline torchtitan/components/lr_scheduler.py]: From what I read, if warmup=4, stable=2, decay=4, train steps=10, what will happen on each step is:
(Here I'm assuming `lr_decay_type == "linear"` and `lr_min=0`)

step 1. lr=0, not learning
step 2. lr=0.25
step 3. lr=0.5
step 4. lr=0.75
step 5. lr=1
step 6. lr=1
step 7. lr=1
step 8. lr=0.75
step 9. lr=0.5
step 10. lr=0.25

This will cause no learning in the first step. Is this expected?

**Review by tianyu-l — COMMENTED** [review]
Glad that you found the bug!

I think the WSD lr scheduler is trickier than it sounds, due to ambiguous definition of warmup/stable/decay steps. See my earlier comment https://github.com/pytorch/torchtitan/pull/1010#pullrequestreview-2718763377

In particular, I believe your change would cause no learning to happen in the first iteration, unless [`lr_min`](https://github.com/pytorch/torchtitan/blob/main/torchtitan/config_manager.py#L173) is specified.

It is OK to me if there is a consensus this should be the default behavior.

File: torchtitan/components/lr_scheduler.py (hunk 2 above)
Comment (CarlosGomes98) [reply torchtitan/components/lr_scheduler.py]: That's a good point, the first step is wasted. I'm not sure what the implementation consensus is, and for some reason its not so easy to find major libraries that have this clearly documented... But the algorithm that was there before now makes sense to me.

I suppose I would propose then to simply not add 1 to the step, only to the warmup_steps.

The other option is to instead enforce a min_lr to be specified, but this is a breaking change

**Comment by wwwjn** [comment]
@CarlosGomes98  I started this https://github.com/pytorch/torchtitan/pull/1284 to make lr_scheduler more clear

**Comment by CarlosGomes98** [comment]
@wwwjn Nice! Then we can close this. Just one caveat, because the lr_scheduler is called after the optimization step takes place, it actually gets called training_step + 1 times. This is what was causing the issue with checkpoint loading.

When we have 10 training steps, and lr_scheduler is called after the last step, I think the correct behaviour is to not change the lr at all in that call. Could you check if your fix deals with this case? Otherwise, an incorrect lr will be stored by the checkpointer

**Comment by wwwjn** [comment]
> @wwwjn Nice! Then we can close this. Just one caveat, because the lr_scheduler is called after the optimization step takes place, it actually gets called training_step + 1 times. This is what was causing the issue with checkpoint loading.
> 
> When we have 10 training steps, and lr_scheduler is called after the last step, I think the correct behaviour is to not change the lr at all in that call. Could you check if your fix deals with this case? Otherwise, an incorrect lr will be stored by the checkpointer

To better describe the issue, the`lr_scheduler.step()` is called after the `optimizer.step()`, and `lr_scheduler.step()` is called `train_step` times, thus we have `train_step + 1` lr states.  When we have 10 total training steps, we actually generate the 11th lr state after the lr_scheduler.step() is called, but the 11th state is not used. 

When the checkpointing is enabled, it will save the 11th lr state into the checkpoint. When resuming the learning rate from the checkpoint, and the optimizer will use the 11th lr state to update weights (by calling `optimzier.step()`). Then we update the lr from 11th state to 12th state (by calling `lr_scheduler.step()`), and the 12th lr state will be used in next iteration. 

In this sense, the checkpoint saves the **next** **lr** that would be used , not the **last lr** that has been used.  I think this behavior is as expected and correct. Wdyt cc @CarlosGomes98
